experimental/tools/handshake-simulator/Simulation.cpp

```diff
@@ -730,120 +284,54 @@ HandshakeExecuter::HandshakeExecuter(
 
-    // Consume the inputs.
-    for (mlir::Value in : op.getOperands())
-      valueMap.erase(in);
-
-    ExecuteStrategy strat = ExecuteStrategy::Default;
-    LogicalResult res =
-        llvm::TypeSwitch<Operation *, LogicalResult>(&op)
-            .Case<mlir::arith::ConstantIndexOp, mlir::arith::ConstantIntOp,
-                  mlir::arith::AddIOp, mlir::arith::AddFOp, mlir::arith::CmpIOp,
-                  mlir::arith::CmpFOp, mlir::arith::SubIOp, mlir::arith::SubFOp,
-                  mlir::arith::MulIOp, mlir::arith::MulFOp,
-                  mlir::arith::DivSIOp, mlir::arith::DivUIOp,
-                  mlir::arith::DivFOp, mlir::arith::IndexCastOp,
-                  mlir::arith::ExtSIOp, mlir::arith::ExtUIOp,
-                  mlir::arith::XOrIOp>(
-                [&](auto op) {
-                  strat = ExecuteStrategy::Default;
-                  return execute(op, inValues, outValues);
-                })
-            .Case<circt::handshake::ReturnOp>([&](auto op) {
-              strat = ExecuteStrategy::Return;
-              return execute(op, inValues, outValues);
-            })
-            .Default([&](auto op) {
-              return op->emitOpError() << "Unknown operation";
-            });
-    LLVM_DEBUG(dbgs() << "EXECUTED: " << op << "\n");
-
-    if (res.failed()) {
-      successFlag = false;
-      return;
-    }
+  // Main simulation loop
+  while (true) {
+    // 1 cycle
+    while (manager.valueChangedThisCycle) {
+      manager.valueChangedThisCycle = false;
+
+      // Note PR : this iterates only over the entry block. I am not sure
+      //          how I should iterate over all operations without having
+      //          4 nested loop, and having an Operation &op reference like this
+      for (Operation &op : entryBlock.getOperations()) {
+        auto opName = op.getName().getStringRef().str();
+        auto &execModel = models[opName];
+
+        if (execModel) {          
+          // Execute operation accordingly to its execution model
+          if (execModel.get()->tryExecute(execData, op)) {
+            // Update the cycle ; a value changed thus we need to keep cycling
+            manager.valueChangedThisCycle = true;
+            if (execModel.get()->isEndPoint()) {
+              successFlag = true;
+              return;
+            }
+          }
 
-    if (strat & ExecuteStrategy::Return)
-      return;
+        } 
+      } // for operations
+    } // while (manager.valueChanged)
 
-    for (auto out : enumerate(op.getResults())) {
-      LLVM_DEBUG(debugArg("OUT", out.value(), outValues[out.index()], time));
-      assert(outValues[out.index()].has_value());
-      valueMap[out.value()] = outValues[out.index()];
-      timeMap[out.value()] = time + 1;
-      scheduleUses(readyList, valueMap, out.value());
-    }
-    ++instructionsExecuted;
-  }
+    // NOTE PR : We might want to have a MAX_CYCLE constant to avoid infinite
```

Comment:
That's a good point. Going even further, I think we want the simulation to scream "deadlock" at the user and stop itself if it detects no changes (even internal operation changes) during a given cycle. That would mean that the circuit cannot do anything anymore.

Note that in reality that may not be true: the circuit might be waiting for some data coming from some BRAM outside the circuit, in which case the circuit may just be waiting for some number of cycles without internally doing anything. However, since we simulate BRAMs and (I assume) they always answer on the next cycle (correct me if I'm wrong) this will not happen for us.